scripts/build_cutout.py

```diff
@@ -92,28 +88,38 @@
 """
 
 import logging
+import atlite
+import geopandas as gpd
+import pandas as pd
 from _helpers import configure_logging
 
-import os
-import atlite
+
+logger = logging.getLogger(__name__)
 
 logger = logging.getLogger(__name__)
 
 if __name__ == "__main__":
     if 'snakemake' not in globals():
         from _helpers import mock_snakemake
-        snakemake = mock_snakemake('build_cutout', cutout='europe-2013-era5')
+        snakemake = mock_snakemake('build_cutout', cutout='base')
     configure_logging(snakemake)
 
     cutout_params = snakemake.config['atlite']['cutouts'][snakemake.wildcards.cutout]
-    for p in ('xs', 'ys', 'years', 'months'):
-        if p in cutout_params:
-            cutout_params[p] = slice(*cutout_params[p])
 
-    cutout = atlite.Cutout(snakemake.wildcards.cutout,
-                        cutout_dir=os.path.dirname(snakemake.output[0]),
-                        **cutout_params)
+    snapshots = pd.date_range(freq='h', **snakemake.config['snapshots'])
+    time = [snapshots[0], snapshots[-1]]
+    cutout_params['time'] = slice(*cutout_params.get('time', time))
+
+    if {'x', 'y', 'bounds'}.isdisjoint(cutout_params):
+        regions = (gpd.read_file(snakemake.input.regions_offshore).append(
+                   gpd.read_file(snakemake.input.regions_onshore)))
+        cutout_params['bounds'] = regions.cascaded_union.buffer(0.5).bounds
+    elif {'x', 'y'}.issubset(cutout_params):
+        cutout_params['x'] = slice(*cutout_params['x'])
+        cutout_params['y'] = slice(*cutout_params['y'])
```

Comment:
So if you specify no bounds, you automatically create a cutout of the bounding box of considered countries? Cool!

Reply:
indeed, also works for non-defined time-spans. So leaving everything blank does everything you need (in standard 0.25 x 0.25 resolution).

Reply:
As with `time` I'd personally favour the script to fail if the bounds are not giving explicitly.
I think we either go we specifying extent and time explicitly or implicitly, but I am against doing it manually (primary) and automatically (fallback).

Rather what we could *always* to, is check whether the bounding boxes are within the cutout extent and - if not - fail.

Reply:
So this does not work with the SARAH dataset then? (IMHO a strong point against the code as it is then)

I guess making it automatic sounds nice, but is it worth the trouble?
Cutouts / cutout config shouldn't have to be adjusted or changed too often, I guess really only once
if you initialise the model or change the geographical scope.

I'm not against fancy solutions and automatisims, just afraid of models doing stuff which becomes hard to debug and people no longer having to give some basic thought to the underlying reanalysis data.

Reply:
No, it also works with the Sarah data. I was reffering to your suggestion 

*Rather what we could always do, is check whether the bounding boxes are within the cutout extent and - if not - fail.*

What I meant is that checking that the bounds of the cutout are withing the bounds of the network does not necessarily play out. The sarah data is only defined until 65 degree north, which is why we set the upper bound to this value. So, there are valid cases where you select an sub-area within your network area which still leads to the same result. 

Nevertheless, the automated bound setting would set the whole covered area. 
I imagine a use case where e.g. you want to make a detailed study on the French grid for the years 2013/2014/2015. All you would have to do is define a new cutout with the module and the rest is done by the workflow. Setting bounding boxes correct is sometimes very annoying and in fact you'd have to do exactly the same what is done in the code.

Reply:
This again leaves me with open questions:
1. Why would you recreate (in your example) the cutout, when you already have our default cutout?
2. If you automatically determine the bounds and they are outside 65°N, what does `atlite` do for the grid cells not covered by SARAH?
3. Why is a `.buffer(0.5)` sufficient? Should it be lower or higher and how is it related to the grid resolution.

Reply:
1. Because you consider different years (not covered by  the default cutout) and you wouldn't want to create a cutout for whole Europe when you just need a single country. 
2. They are effectively set to zero. This happens in the conversion to pv with a `.fillna(0)`
3. good point, it should probably be the max of dx and dy

Reply:
1. Makes sense.
2. Ok, fair enough.

I'm still not convinced we need this feature and not a big fan of it. But from a code-quality I think it is ok from my POV.
(One or two comments in the code would be a nice ;)